Properties to set default configuration for auto-timed controller metrics

ttddyy opened the pull request:
When `management.metrics.web.server.autoTimeRequests` is enabled (default=true), spring-boot collects metrics on controller methods even when they are not annotated with `@Timed`.
When this happens, created metrics are based on the default configuration values of `@Timed`.
Currently, there is no way to specify the default configuration to those auto-timed controller metrics.

This PR introduces two properties:
- `management.metrics.web.server.autoTimeRequestsDefaultPercentiles`
- `management.metrics.web.server.autoTimeRequestsDefaultHistogram`

When values are set to the above properties, they will be applied to auto-timed controller metrics that do not have explicit `@Timed` annotation.

wilkinsona commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -133,6 +135,20 @@ public void setMaxUriTags(int maxUriTags) {
+			 * {@link Timed} is not presented on the corresponding request handler. Any
+			 * {@link Timed} annotation presented will have precedence.
+			 */
+			private double[] autoTimeRequestsDefaultPercentiles;
```
The field javadoc is used to document the configuration property in the reference documentation appendix and in the metadata that's used by IDEs. As such, it can't use javadoc tags and must be written in plain text. Can you please rework the description to match the style of the other fields in the class?

wilkinsona commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -133,6 +135,20 @@ public void setMaxUriTags(int maxUriTags) {
+			 * {@link Timed} is not presented on the corresponding request handler. Any
+			 * {@link Timed} annotation presented will have precedence.
+			 */
+			private boolean autoTimeRequestsDefaultHistogram;
```
The field javadoc is used to document the configuration property in the reference documentation appendix and in the metadata that's used by IDEs. As such, it can't use javadoc tags and must be written in plain text. Can you please rework the description to match the style of the other fields in the class?

wilkinsona commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/servlet/WebMvcMetricsFilter.java:
```diff
@@ -88,10 +92,31 @@ public WebMvcMetricsFilter(ApplicationContext context, MeterRegistry registry,
 	 */
 	public WebMvcMetricsFilter(MeterRegistry registry, WebMvcTagsProvider tagsProvider,
```
I think we should deprecate this constructor in favour of the new one added below.

wilkinsona commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/reactive/server/MetricsWebFilter.java:
```diff
@@ -64,10 +69,19 @@ public MetricsWebFilter(MeterRegistry registry, WebFluxTagsProvider tagsProvider
 
 	public MetricsWebFilter(MeterRegistry registry, WebFluxTagsProvider tagsProvider,
```
I think we should deprecate this constructor in favour of the new one added below.

ttddyy commented:
Hi @wilkinsona 
Updated the PR.
I was not sure for what version to put in deprecation javadoc, so I put `2.1.4`.

Thanks,

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -133,6 +133,20 @@ public void setMaxUriTags(int maxUriTags) {
+			 * not presented on the corresponding request handler. Any @Timed annotation
+			 * presented will have precedence.
+			 */
+			private double[] autoTimeRequestsDefaultPercentiles;
```
I am not sure about the `double[]` array, I'd rather use a `List` for this.

Also we now have three properties around the concept of `autoTimeRequests` and some keys are now rather long. Shouldn't we take the opportunity to group those in a group?

snicoll commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/reactive/server/MetricsWebFilter.java:
```diff
@@ -48,12 +49,46 @@
+	 * @param tagsProvider provider for metrics tags
+	 * @param metricName name of the metric to record
+	 * @param autoTimeRequests if requests should be automatically timed
+	 * @deprecated since 2.1.4 in favor of
```
This PR is intended to be merged on `master` only so that should be deprecated as of `2.2.0`

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -133,6 +133,20 @@ public void setMaxUriTags(int maxUriTags) {
 			private boolean autoTimeRequests = true;
 
+			/**
+			 * Default percentiles when "autoTimeRequests=true" and @Timed annotation is
```
I don't like the idea of referring one property from another property. The canonical format is not camelCase anyway.

ttddyy commented:
Added a commit that moves auto-time-requests related properties to an independent `AutoTimeRequests` properties class.

wilkinsona commented:
We discussed this a bit and wondered if properties like `management.metrics.web.server.request.autotime.default-histogram` would be preferable. We'd have similar properties on the client-side beneath `management.metrics.web.client.request`. We'd need to deprecate some existing properties such as `management.metrics.web.server.requests-metric-name` and move it to something like `management.metrics.web.server.request.metric-name`

ttddyy commented:
@wilkinsona 

Updated the review for server side auto-time metrics.

Updated auto-time requests related config to:
- `management.metrics.web.server.request.auto-time.enabled`
- `management.metrics.web.server.request.auto-time.default-percentiles`
- `management.metrics.web.server.request.auto-time.default-histogram`

Updated `requests-metric-name`:
`management.metrics.web.server.requests-metric-name`
=> `management.metrics.web.server.request.metric-name`

Not sure for how to deprecate the entry on properties class.
For now, I just put `@deprecated` for `management.metrics.web.server.requests-metric-name`.
If it is ok to just remove them, let me know.

I'll find some time for doing client-side.

ttddyy commented:
ok, updated the client side as well.

ttddyy commented:
@wilkinsona  
I have incorporated the comments and force pushed the squashed commit.
Please take a look.

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -93,10 +96,7 @@ public Server getServer() {
-			 * Name of the metric for sent requests.
-			 */
-			private String requestsMetricName = "http.client.requests";
+			private final Request request = new Request("http.client.requests");
```
Unfortunately that's not very consistent with what we want in the metadata. The default should be in code rather. Let's go with a `ClientRequest` that hardcodes the value as it used to in the field.

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -105,12 +105,26 @@ public Server getServer() {
+			 * Get name of the metric for received requests.
+			 * @return request metric name
+			 * @deprecated since 2.2.0 in favor of {@link Request#getMetricName()}
+			 */
```
This should have a `@DeprecatedConfigurationProperty` with the replacement key

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -125,18 +139,7 @@ public void setMaxUriTags(int maxUriTags) {
-			 * Name of the metric for received requests.
-			 */
-			private String requestsMetricName = "http.server.requests";
+			private final Request request = new Request("http.server.requests");
```
Idem as above with a `ServerRequest`

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -145,20 +148,26 @@ public void setMaxUriTags(int maxUriTags) {
+			 * Get name of the metric for received requests.
+			 * @return request metric name
+			 * @deprecated since 2.2.0 in favor of {@link Request#getMetricName()}
+			 */
```
This should have a @DeprecatedConfigurationProperty with the replacement key

snicoll commented on spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MetricsProperties.java:
```diff
@@ -171,6 +180,86 @@ public void setMaxUriTags(int maxUriTags) {
 
+		public static class Request {
+
+			public Request(String metricName) {
```
With 2.2 this will break as the binding will consider you want to use constructor binding with a single property. Please remove that constructor.

snicoll commented on spring-boot-project/spring-boot-docs/src/main/asciidoc/production-ready-features.adoc:
```diff
@@ -1801,7 +1801,7 @@ application's absolute start time
 ==== Spring MVC Metrics
 Auto-configuration enables the instrumentation of requests handled by Spring MVC. When
-`management.metrics.web.server.auto-time-requests` is `true`, this instrumentation occurs
+`management.metrics.web.server.auto-time-requests.enabled` is `true`, this instrumentation occurs
```
That doesn't look like the right key. `request.auto-time.enabled` ?

snicoll commented on spring-boot-project/spring-boot-docs/src/main/asciidoc/production-ready-features.adoc:
```diff
@@ -1896,7 +1896,7 @@ To customize the tags, provide a `@Bean` that implements `WebFluxTagsProvider`.
 ==== Jersey Server Metrics
 Auto-configuration enables the instrumentation of requests handled by the Jersey JAX-RS
-implementation. When `management.metrics.web.server.auto-time-requests` is `true`, this
+implementation. When `management.metrics.web.server.auto-time-requests.enabled` is `true`, this
```
That doesn't look like the right key. `request.auto-time.enabled` ?

ttddyy commented:
ok, incorporated the review comments and updated/rebased/squashed the commit.

snicoll commented:
@ttddyy thank you for your hard work on this one. This is merged on `master` with [a polish commit](https://github.com/spring-projects/spring-boot/commit/8045bf1f50ba7e41552d5e3b585c9ebca39eecd3).

snicoll commented:
I can't reopen this pull request but the `auto-time-requests` property has been removed rather than being deprecated and I missed that as part of the polish. I'll try to restore it in a deprecated fashion now.
